cpp/src/strings/convert/convert_urls.cu

```diff
@@ -185,137 +193,197 @@ constexpr bool is_hex_digit(char ch)
   }
-};
+}
 
 /**
- * @brief Functor to update the string column offsets.
+ * @brief Decode and copy from the input char column to the output char column.
+ *
+ * @tparam num_warps_per_threadblock Number of warps in a threadblock. This template argument must
+ * match the launch configuration, i.e. the kernel must be launched with
+ * `num_warps_per_threadblock * cudf::detail::warp_size` threads per threadblock.
+ * @tparam char_block_size Number of characters which will be loaded into the shared memory at a
+ * time.
+ *
+ * @param[in] in_chars Character buffer for the input string column.
+ * @param[in] in_offsets Offset value of each string associated with `in_chars`.
+ * @param[out] out_chars Character buffer for the output string column.
+ * @param[in] out_offsets Offset value of each string associated with `out_chars`.
+ * @param[in] num_strings Number of strings to decode and copy.
  */
-struct url_decode_offsets_updater {
-  device_span<size_type const> const d_esc_positions{};
-  int32_t const first_input_offset = 0;
-
-  /**
-   * @brief Convert input offsets into output offsets
-   *
-   * Each offset is reduced by 2 for every escape sequence that occurs in the entire string column
-   * character data before the offset, as 3 characters are replaced with 1 for each escape.
-   *
-   * @param offset An original offset value from the input string column
-   * @return Adjusted offset value
-   */
-  __device__ int32_t operator()(int32_t offset) const
-  {
-    // determine the number of escape sequences occurring before this offset
-    size_type const* next_esc_pos_ptr =
-      thrust::lower_bound(thrust::seq, d_esc_positions.begin(), d_esc_positions.end(), offset);
-    size_type num_prev_esc = next_esc_pos_ptr - d_esc_positions.data();
-    // every escape that occurs before this one replaces 3 characters with 1
-    return offset - first_input_offset - (num_prev_esc * 2);
+template <int num_warps_per_threadblock, int char_block_size>
+__global__ void url_decode_char_replacer(char const* const in_chars,
+                                         offset_type const* const in_offsets,
+                                         char* const out_chars,
+                                         offset_type const* const out_offsets,
+                                         size_type const num_strings)
+{
+  constexpr int halo_size = 2;
+  __shared__ char temporary_buffer[num_warps_per_threadblock][char_block_size + halo_size * 2];
+  __shared__ typename cub::WarpScan<int8_t>::TempStorage cub_storage[num_warps_per_threadblock];
+  __shared__ int out_idx[num_warps_per_threadblock];
+
+  int global_thread_id  = blockIdx.x * blockDim.x + threadIdx.x;
+  int global_warp_id    = global_thread_id / cudf::detail::warp_size;
+  int local_warp_id     = threadIdx.x / cudf::detail::warp_size;
+  int warp_lane         = threadIdx.x % cudf::detail::warp_size;
+  int nwarps            = gridDim.x * blockDim.x / cudf::detail::warp_size;
```

Comment:
```suggestion
  int const global_thread_id  = blockIdx.x * blockDim.x + threadIdx.x;
  int const global_warp_id    = global_thread_id / cudf::detail::warp_size;
  int const local_warp_id     = threadIdx.x / cudf::detail::warp_size;
  int const warp_lane         = threadIdx.x % cudf::detail::warp_size;
  int const nwarps            = gridDim.x * blockDim.x / cudf::detail::warp_size;
```

Reply:
Addressed in https://github.com/rapidsai/cudf/pull/8622/commits/6a671bc4981b3e21c170440eafafb7e9b163f2b9.